pkg/operator/controller/controller_dns_node_resolver_daemonset.go

```diff
@@ -0,0 +1,264 @@
+package controller
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+
+	"github.com/openshift/cluster-dns-operator/pkg/manifests"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+const (
+	// services is a comma- or space-delimited list of services for which
+	// entries should be added to /etc/hosts.  NOTE: For now, ensure these
+	// are relative names; for each relative name, an alias with the
+	// CLUSTER_DOMAIN suffix will also be added.
+	services = "image-registry.openshift-image-registry.svc"
+)
+
+var (
+	// nodeResolverScript is a shell script that updates /etc/hosts.
+	nodeResolverScript = manifests.NodeResolverScript()
+)
+
+// ensureNodeResolverDaemonset ensures the node resolver daemonset exists if it
+// should or does not exist if it should not exist.  Returns a Boolean
+// indicating whether the daemonset exists, the daemonset if it does exist, and
+// an error value.
+func (r *reconciler) ensureNodeResolverDaemonSet(clusterIP, clusterDomain string) (bool, *appsv1.DaemonSet, error) {
+	haveDS, current, err := r.currentNodeResolverDaemonSet()
+	if err != nil {
+		return false, nil, err
+	}
+	wantDS, desired, err := desiredNodeResolverDaemonSet(clusterIP, clusterDomain, r.OpenshiftCLIImage)
+	if err != nil {
+		return haveDS, current, fmt.Errorf("failed to build node resolver daemonset: %v", err)
+	}
+	switch {
+	case !wantDS && !haveDS:
+		return false, nil, nil
+	case !wantDS && haveDS:
+		if err := r.deleteNodeResolverDaemonSet(current); err != nil {
+			return true, current, err
+		}
+		return false, nil, nil
+	case wantDS && !haveDS:
+		if err := r.createNodeResolverDaemonSet(desired); err != nil {
+			return false, nil, err
+		}
+		return r.currentNodeResolverDaemonSet()
+	case wantDS && haveDS:
+		if updated, err := r.updateNodeResolverDaemonSet(current, desired); err != nil {
+			return true, current, err
+		} else if updated {
+			return r.currentNodeResolverDaemonSet()
+		}
+	}
+	return true, current, nil
+}
+
+// desiredNodeResolverDaemonSet returns the desired node resolver daemonset.
+func desiredNodeResolverDaemonSet(clusterIP, clusterDomain, openshiftCLIImage string) (bool, *appsv1.DaemonSet, error) {
+	hostPathFile := corev1.HostPathFile
+	maxUnavailable := intstr.FromString("10%")
+	envs := []corev1.EnvVar{{
+		Name:  "SERVICES",
+		Value: services,
+	}}
+	if len(clusterIP) > 0 {
+		envs = append(envs, corev1.EnvVar{
+			Name:  "NAMESERVER",
+			Value: clusterIP,
+		})
+	}
+	if len(clusterDomain) > 0 {
+		envs = append(envs, corev1.EnvVar{
+			Name:  "CLUSTER_DOMAIN",
+			Value: clusterDomain,
+		})
+	}
+	trueVal := true
+	name := NodeResolverDaemonSetName()
+	daemonset := appsv1.DaemonSet{
```

Comment:
I am not opposed to defining this new daemonset completely in code, but out of curiosity, any particular reason not to make the daemonset skeleton available via `assets/`?

Reply:
I'm kind of fed up with yaml:

* Running `make generate` after updating yaml is an extra step for developers.  
* Loading the yaml so that we can mutate the result is an extra step in the code.  
* Modifying `bindata.go` leads to more merge conflicts, especially during backports, and thus more work for developers.  
* Using yaml instead of Go leads to uncaught syntax errors, such as the one that https://github.com/openshift/cluster-dns-operator/pull/221 fixes.  (Maybe we just need better unit tests.)  

Thus I'm disinclined to add new yaml assets for values that we manipulate in Go code.  

As far as I'm concerned, the only situations in which using assets is making life easier for us are simple yaml that we don't manipulate in Go code, such as CVO manifests and RBAC resources; and assets that are useful independently, such as the `update-node-resolver.sh` shell script that you might want to run directly or assets that you are likely to want to pass into `oc apply` or the like.

Reply:
> As far as I'm concerned, the only situations in which using assets is making life easier for us are simple yaml that we don't manipulate in Go code, such as CVO manifests and RBAC resources; and assets that are useful independently, such as the update-node-resolver.sh shell script that you might want to run directly or assets that you are likely to want to pass into oc apply or the like.

Sounds good to me.